diff --git a/src/drivers/fluid_jack.c b/src/drivers/fluid_jack.c
index fe3cd75..613a982 100644
--- a/src/drivers/fluid_jack.c
+++ b/src/drivers/fluid_jack.c
@@ -100,7 +100,7 @@ int fluid_jack_driver_process(jack_nframes_t nframes, void *arg);
 int delete_fluid_jack_midi_driver(fluid_midi_driver_t *p);
 
 
-static fluid_mutex_t last_client_mutex = G_STATIC_MUTEX_INIT;     /* Probably not necessary, but just in case drivers are created by multiple threads */
+static fluid_mutex_t last_client_mutex;     /* Probably not necessary, but just in case drivers are created by multiple threads */
 static fluid_jack_client_t *last_client = NULL;       /* Last unpaired client. For audio/MIDI driver pairing. */
 
 
